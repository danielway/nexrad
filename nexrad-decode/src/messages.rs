pub mod clutter_filter_map;
pub mod digital_radar_data;
pub mod message_header;
pub mod rda_status_data;

mod definitions;
mod primitive_aliases;

mod message_type;
pub use message_type::MessageType;

mod message;
pub use message::{Message, MessageWithHeader};
