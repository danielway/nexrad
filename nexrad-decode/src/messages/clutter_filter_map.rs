//!
//! Message type 15 "Clutter Filter Map" contains information about clutter filter maps that are
//! used to filter clutter from radar products. The clutter filter map is a 3D array of elevation,
//! azimuth, and range zones that define the clutter filter behavior for radar products.
//!

mod header;
pub use header::Header;

mod message;
pub use message::Message;

mod elevation_segment;
pub use elevation_segment::ElevationSegment;

mod azimuth_segment;
pub use azimuth_segment::{AzimuthSegment, AzimuthSegmentHeader};

mod range_zone;
pub use range_zone::RangeZone;

mod definitions;
pub use definitions::*;

use crate::result::Result;
use crate::util::deserialize;
use std::io::Read;

/// Decodes a clutter filter map message type 15 from the provided reader.
pub fn decode_clutter_filter_map<R: Read>(reader: &mut R) -> Result<Message> {
    let header: Header = deserialize(reader)?;
    let elevation_segment_count = header.elevation_segment_count as u8;

    let mut message = Message::new(header);

    for elevation_segment_number in 0..elevation_segment_count {
        let mut elevation_segment = ElevationSegment::new(elevation_segment_number);

        for azimuth_number in 0..360 {
            let azimuth_segment_header: AzimuthSegmentHeader = deserialize(reader)?;
            let range_zone_count = azimuth_segment_header.range_zone_count as usize;

            let mut azimuth_segment = AzimuthSegment::new(azimuth_segment_header, azimuth_number);
            for _ in 0..range_zone_count {
                azimuth_segment.range_zones.push(deserialize(reader)?);
            }

            elevation_segment.azimuth_segments.push(azimuth_segment);
        }

        message.elevation_segments.push(elevation_segment);
    }

    Ok(message)
}
