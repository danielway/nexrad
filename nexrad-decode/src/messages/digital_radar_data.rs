//!
//! Message type 31 "Digital Radar Data" consists of base data information such as reflectivity,
//! mean radial velocity, spectrum width, differential reflectivity, differential phase, correlation
//! coefficient, azimuth angle, elevation angle, cut type, scanning strategy, and calibration
//! parameters. The frequency and volume of the message is dependent on the scanning strategy and
//! the type of data associated with that strategy.
//!

mod header;
pub use header::Header;

mod message;
pub use message::Message;

mod data_block_id;
pub use data_block_id::DataBlockId;

mod volume_data_block;
pub use volume_data_block::VolumeDataBlock;

mod generic_data_block;
pub use generic_data_block::{GenericDataBlock, GenericDataBlockHeader};

mod elevation_data_block;
pub use elevation_data_block::ElevationDataBlock;

mod radial_data_block;
pub use radial_data_block::RadialDataBlock;

mod definitions;
pub use definitions::*;

mod spot_blanking_status;
pub use spot_blanking_status::*;

mod pointers;
pub use pointers::*;
