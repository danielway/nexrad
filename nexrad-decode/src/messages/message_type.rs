/// The types of data messages transferred between the RDA and RPG.
#[derive(Debug, Hash, PartialEq, Eq, Clone, Copy, Ord, PartialOrd)]
pub enum MessageType {
    /// Replaced by message type 31.
    RDADigitalRadarData = 1,

    /// Metadata.
    RDAStatusData = 2,

    /// Metadata.
    RDAPerformanceMaintenanceData = 3,

    RDAConsoleMessage = 4,

    /// Metadata.
    RDAVolumeCoveragePattern = 5,

    RDAControlCommands = 6,

    RPGVolumeCoveragePattern = 7,

    RPGClutterCensorZones = 8,

    RPGRequestForData = 9,

    RPGConsoleMessage = 10,

    RDALoopBackTest = 11,

    RPGLoopBackTest = 12,

    /// No longer sent.
    RDAClutterFilterBypassMap = 13,

    Spare1 = 14,

    /// Metadata.
    RDAClutterFilterMap = 15,

    ReservedFAARMSOnly1 = 16,

    ReservedFAARMSOnly2 = 17,

    /// Metadata.
    RDAAdaptationData = 18,

    Reserved1 = 20,

    Reserved2 = 21,

    Reserved3 = 22,

    Reserved4 = 23,

    ReservedFAARMSOnly3 = 24,

    ReservedFAARMSOnly4 = 25,

    ReservedFAARMSOnly5 = 26,

    Reserved5 = 29,

    RDADigitalRadarDataGenericFormat = 31,

    RDAPRFData = 32,

    RDALogData = 33,

    Unknown,
}
