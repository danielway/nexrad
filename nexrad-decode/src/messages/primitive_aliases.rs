//!
//! Primitive aliases matching the types referenced in the ICD.
//!

use zerocopy::big_endian;

pub type Code1 = u8;
pub type Code2 = big_endian::U16;
pub type Integer1 = u8;
pub type Integer2 = big_endian::U16;
pub type Integer4 = big_endian::U32;
pub type Real4 = big_endian::F32;
pub type ScaledInteger1 = u8;
pub type ScaledInteger2 = big_endian::U16;
pub type ScaledSInteger2 = big_endian::I16;
pub type SInteger2 = big_endian::I16;
