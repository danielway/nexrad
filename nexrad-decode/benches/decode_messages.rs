use std::{hint::black_box, io::Cursor, time::Duration};

use criterion::{criterion_group, criterion_main, BatchSize, Criterion};
use nexrad_data::volume;
use nexrad_decode::messages::decode_messages;

const TEST_NEXRAD_FILE: &[u8] = include_bytes!("../../downloads/KDMX20220305_232324_V06");

fn benchmark_decode_messages(c: &mut Criterion) {
    let volume = volume::File::new(TEST_NEXRAD_FILE.to_vec());
    let mut records = volume.records().into_iter().take(2);

    let mut first_record = records.next().expect("first record exists");
    first_record = first_record
        .decompress()
        .expect("decompresses first record");

    let record1_data = first_record.data().to_vec();

    let mut second_record = records.next().expect("second record exists");
    if second_record.compressed() {
        second_record = second_record
            .decompress()
            .expect("decompresses second record");
    }

    let record2_data = second_record.data().to_vec();

    let mut group = c.benchmark_group("decode_messages");
    group
        .warm_up_time(Duration::from_secs(5))
        .measurement_time(Duration::from_secs(15))
        .sample_size(200)
        .noise_threshold(0.05)
        .significance_level(0.02);

    group.bench_function("record_0", |b| {
        b.iter_batched(
            || record1_data.clone(),
            |data| {
                let mut reader = Cursor::new(black_box(&data));
                decode_messages(&mut reader).expect("decodes successfully")
            },
            BatchSize::SmallInput,
        )
    });

    group.bench_function("record_1", |b| {
        b.iter_batched(
            || record2_data.clone(),
            |data| {
                let mut reader = Cursor::new(black_box(&data));
                decode_messages(&mut reader).expect("decodes successfully")
            },
            BatchSize::SmallInput,
        )
    });

    group.finish();
}

criterion_group!(benches, benchmark_decode_messages);
criterion_main!(benches);
