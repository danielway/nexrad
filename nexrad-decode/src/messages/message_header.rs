use crate::binary_data::BinaryData;
use crate::messages::definitions::RedundantChannel;
use crate::messages::message_type::MessageType;
use crate::messages::primitive_aliases::{Integer1, Integer2, Integer4};
use crate::util::get_datetime;
use chrono::{DateTime, Duration, Utc};
use std::fmt::Debug;
use zerocopy::{TryFromBytes, Immutable, KnownLayout};

#[cfg(feature = "uom")]
use uom::si::f64::Information;
#[cfg(feature = "uom")]
use uom::si::information::byte;

/// This value in the [MessageHeader::segment_size] field of a message header indicates that the
/// message is variable-length rather than segmented.
pub const VARIABLE_LENGTH_MESSAGE_SIZE: u16 = 65535;

/// Message and system configuration information appended to the beginning of all messages.
///
/// Note that messages with a segment size of [VARIABLE_LENGTH_MESSAGE_SIZE] are not segmented and
/// instead variable-length, with the segment count and segment number positions of the header
/// (bytes 12-15) specifying the size of the full message in bytes.
#[repr(C)]
#[derive(Clone, PartialEq, Eq, Hash, Debug, TryFromBytes, Immutable, KnownLayout)]
pub struct MessageHeader {
    /// Unknown/reserved bytes from RPG.
    pub rpg_unknown: BinaryData<[u8; 12]>,

    /// Size of this segment in half-words. Note that this only describes this segment's size,
    /// though there could be multiple segments. In the case of a variable-length message (indicated
    /// by this field being set to [VARIABLE_LENGTH_MESSAGE_SIZE]), the full message's size is
    /// determined differently. See [MessageHeader::message_size_bytes] and
    /// [MessageHeader::segment_count] for more information.
    pub segment_size: Integer2,

    /// Whether the RDA is operating on a redundant channel.
    ///
    /// Legacy:
    ///  0 = Single Channel (no bits set)
    ///  1 = Redundant Channel 1 (bit 0 set)
    ///  2 = Redundant Channel 2 (bit 1 set)
    /// ORDA
    ///  8 = Single Channel (bit 3 set)
    ///  9 = Redundant Channel 1 (bits 0 and 3 set)
    /// 10 = Redundant Channel 2 (bits 1 and 3 set)
    pub redundant_channel: Integer1,

    /// Type discriminator.
    pub message_type: Integer1,

    /// Message sequence number.
    pub sequence_number: Integer2,

    /// This message's date represented as a count of days since 1 January 1970 00:00 GMT. It is
    /// also referred-to as a "modified Julian date" where it is the Julian date - 2440586.5.
    pub date: Integer2,

    /// Milliseconds past midnight, GMT.
    pub time: Integer4,

    /// Number of segments in this message. If the [MessageHeader::segment_size] is less than
    /// [VARIABLE_LENGTH_MESSAGE_SIZE], this field is meaningful, otherwise bytes 12-15 (this field
    /// and [MessageHeader::segment_number]) specify the size of the message in bytes.
    pub segment_count: Integer2,

    /// This message segment's number. If the [MessageHeader::segment_size] is less than
    /// [VARIABLE_LENGTH_MESSAGE_SIZE], this field is meaningful, otherwise, bytes 12-15 (this field
    /// and [MessageHeader::segment_number]) specify the size of the message in bytes.
    pub segment_number: Integer2,
}

impl MessageHeader {
    /// If this message is [MessageHeader::segmented], this indicates this message segment's size.
    /// Otherwise, this returns [None] and [MessageHeader::message_size] should be used to determine
    /// the message's full size.
    #[cfg(feature = "uom")]
    pub fn segment_size(&self) -> Option<Information> {
        let segment_size = self.segment_size.get();
        if segment_size < VARIABLE_LENGTH_MESSAGE_SIZE {
            Some(Information::new::<byte>((segment_size * 2) as f64))
        } else {
            None
        }
    }

    /// Whether the RDA is operating on a redundant channel.
    pub fn rda_redundant_channel(&self) -> RedundantChannel {
        match self.redundant_channel {
            0 => RedundantChannel::LegacySingleChannel,
            1 => RedundantChannel::LegacyRedundantChannel1,
            2 => RedundantChannel::LegacyRedundantChannel2,
            8 => RedundantChannel::ORDASingleChannel,
            9 => RedundantChannel::ORDARedundantChannel1,
            10 => RedundantChannel::ORDARedundantChannel2,
            _ => panic!("Invalid RDA redundant channel: {}", self.redundant_channel),
        }
    }

    /// Message type discriminator.
    pub fn message_type(&self) -> MessageType {
        match self.message_type {
            1 => MessageType::RDADigitalRadarData,
            2 => MessageType::RDAStatusData,
            3 => MessageType::RDAPerformanceMaintenanceData,
            4 => MessageType::RDAConsoleMessage,
            5 => MessageType::RDAVolumeCoveragePattern,
            6 => MessageType::RDAControlCommands,
            7 => MessageType::RPGVolumeCoveragePattern,
            8 => MessageType::RPGClutterCensorZones,
            9 => MessageType::RPGRequestForData,
            10 => MessageType::RPGConsoleMessage,
            11 => MessageType::RDALoopBackTest,
            12 => MessageType::RPGLoopBackTest,
            13 => MessageType::RDAClutterFilterBypassMap,
            14 => MessageType::Spare1,
            15 => MessageType::RDAClutterFilterMap,
            16 => MessageType::ReservedFAARMSOnly1,
            17 => MessageType::ReservedFAARMSOnly2,
            18 => MessageType::RDAAdaptationData,
            20 => MessageType::Reserved1,
            21 => MessageType::Reserved2,
            22 => MessageType::Reserved3,
            23 => MessageType::Reserved4,
            24 => MessageType::ReservedFAARMSOnly3,
            25 => MessageType::ReservedFAARMSOnly4,
            26 => MessageType::ReservedFAARMSOnly5,
            29 => MessageType::Reserved5,
            31 => MessageType::RDADigitalRadarDataGenericFormat,
            32 => MessageType::RDAPRFData,
            33 => MessageType::RDALogData,
            _ => MessageType::Unknown(self.message_type),
        }
    }

    /// This message's date and time in UTC.
    pub fn date_time(&self) -> Option<DateTime<Utc>> {
        get_datetime(self.date.get(), Duration::milliseconds(self.time.get() as i64))
    }

    /// Whether this message is segmented or variable-length. If the message is segmented, multiple
    /// message segments compose the full message. If the message is variable-length as indicated by
    /// the [MessageHeader::segment_size] field being set to [VARIABLE_LENGTH_MESSAGE_SIZE], the
    /// full message size can be retrieved by [MessageHeader::message_size_bytes].
    pub fn segmented(&self) -> bool {
        self.segment_size.get() < VARIABLE_LENGTH_MESSAGE_SIZE
    }

    /// If the message is [MessageHeader::segmented], this indicates the number of segments in the
    /// full message, otherwise this returns [None]. [MessageHeader::message_size_bytes] can be used
    /// to determine the message's full size.
    pub fn segment_count(&self) -> Option<u16> {
        if self.segment_size.get() < VARIABLE_LENGTH_MESSAGE_SIZE {
            Some(self.segment_count.get())
        } else {
            None
        }
    }

    /// If the message is [MessageHeader::segmented], this indicates this segment's number/sequence
    /// in the message, otherwise this returns [None]. [MessageHeader::message_size_bytes] can be
    /// used to determine the message's full size.
    pub fn segment_number(&self) -> Option<u16> {
        if self.segment_size.get() < VARIABLE_LENGTH_MESSAGE_SIZE {
            Some(self.segment_number.get())
        } else {
            None
        }
    }

    /// The full size of the message in bytes. If the message is [MessageHeader::segmented] then
    /// this is the segment size, otherwise this is the full variable-length message size.
    pub fn message_size_bytes(&self) -> u32 {
        match self.segment_count() {
            Some(_) => self.segment_size.get() as u32 * 2,
            None => {
                let segment_number = self.segment_number.get() as u32;
                let segment_size = self.segment_size.get() as u32;
                (segment_number << 16) | (segment_size << 1)
            }
        }
    }

    /// The full size of the message. If the message is [MessageHeader::segmented] then this is the
    /// segment size, otherwise this is the full variable-length message size.
    #[cfg(feature = "uom")]
    pub fn message_size(&self) -> Information {
        match self.segment_count() {
            Some(_) => {
                let segment_size_bytes = self.segment_size.get() << 1;
                Information::new::<byte>(segment_size_bytes as f64)
            }
            None => {
                let segment_number = self.segment_number.get() as u32;
                let segment_size = self.segment_size.get() as u32;
                let message_size_bytes = (segment_number << 16) | segment_size;
                Information::new::<byte>(message_size_bytes as f64)
            }
        }
    }

    /// Decodes a reference to a MessageHeader from a byte slice, returning the header and remaining bytes.
    pub fn decode_ref(bytes: &[u8]) -> crate::result::Result<(&Self, &[u8])> {
        Ok(Self::try_ref_from_prefix(bytes)?)
    }

    /// Decodes an owned copy of a MessageHeader from a byte slice.
    pub fn decode_owned(bytes: &[u8]) -> crate::result::Result<Self> {
        let (header, _) = Self::decode_ref(bytes)?;
        Ok(header.clone())
    }
}
