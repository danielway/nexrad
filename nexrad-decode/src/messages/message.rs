use crate::messages::clutter_filter_map;
use crate::messages::digital_radar_data;
use crate::messages::message_header::MessageHeader;
use crate::messages::rda_status_data;

/// A decoded NEXRAD Level II message with its metadata header.
#[derive(Debug)]
pub struct MessageWithHeader {
    pub header: MessageHeader,
    pub message: Message,
}

/// A decoded NEXRAD Level II message.
#[derive(Debug)]
pub enum Message {
    RDAStatusData(Box<rda_status_data::Message>),
    DigitalRadarData(Box<digital_radar_data::Message>),
    ClutterFilterMap(Box<clutter_filter_map::Message>),
    Other,
}
