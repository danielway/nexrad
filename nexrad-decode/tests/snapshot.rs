use std::io::Cursor;

use insta::{assert_yaml_snapshot, internals::Content};
use nexrad_data::volume;
use nexrad_decode::messages::decode_messages;
use sha2::{Digest, Sha256};

const TEST_NEXRAD_FILE: &[u8] = include_bytes!("KDMX20220305_232324_V06");

#[test]
fn test_decode_volume() {
    let volume = volume::File::new(TEST_NEXRAD_FILE.to_vec());
    for (record_number, mut record) in volume.records().into_iter().enumerate() {
        if record.compressed() {
            record = record.decompress().expect("decompresses records");
        }

        let mut reader = Cursor::new(record.data());
        let messages = decode_messages(&mut reader).expect("decodes successfully");

        assert_yaml_snapshot!(format!("Record {}", record_number), messages, {
          ".**.rpg_unknown" => insta::dynamic_redaction(|value, _path| redact_bytes_like(value)),
          ".**.encoded_data" => insta::dynamic_redaction(|value, _path| redact_bytes_like(value)),
        });
    }
}

fn redact_bytes_like(value: insta::internals::Content) -> insta::internals::Content {
    if let Some(items) = value.as_slice() {
        let mut buf = Vec::with_capacity(items.len());
        for item in items {
            if let Some(n) = item.as_u64() {
                if n <= 255 {
                    buf.push(n as u8);
                    continue;
                }
            }
            panic!("unexpected value in binary data: {:?}", item);
        }
        return hash_bytes_to_content(&buf);
    }
    value
}

fn hash_bytes_to_content(bytes: &[u8]) -> insta::internals::Content {
    let mut hash = Sha256::new();
    hash.update(bytes);

    let digest = hex::encode(hash.finalize());
    format!("bytes::<{}>::sha256:{digest}", bytes.len()).into()
}
