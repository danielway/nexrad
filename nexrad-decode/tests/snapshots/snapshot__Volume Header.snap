---
source: nexrad-decode/tests/snapshot.rs
expression: volume.header()
---
Ok(
    Header {
        tape_filename: [
            65,
            82,
            50,
            86,
            48,
            48,
            48,
            54,
            46,
        ],
        extension_number: [
            56,
            55,
            57,
        ],
        date: U32(
            19057,
        ),
        time: U32(
            84204299,
        ),
        icao_of_radar: [
            75,
            68,
            77,
            88,
        ],
    },
)
